### PR TITLE
Add comma-separated --auto-accept support.

### PR DESCRIPTION
**Pull request by johnharris85**
Addresses #24143, simplifying `--auto-accept` syntax for a better UX.

Add support for comma-separated --auto-accept value.

`$ docker swarm update --auto-accept worker,manager`.

Signed-off-by: John Harris john@johnharris.io

### CODE REVIEW

**Comment by dnephin** [comment]
If we're going to support `all` instead of `--auto-accept worker  --auto-accept manager` I think we should make a much larger change to the `AutoAcceptOption`. 

The logic of `Set` becomes a lot simpler if you can't set it more than once.

**Comment by johnharris85** [comment]
I'm implementing the `--auto-accept-add` and `--auto-accept-rm` flags, but wondering how auto-acceptance should be handled on `docker swarm init`. Obviously `add` still works there but should `rm` error? Or just silently do nothing (as it looks like default behavior for init is going to be auto-accept = false for both anyway)?

**Comment by dnephin** [comment]
I guess on `init` it should actually just use the current behaviour

**Comment by thaJeztah** [comment]
I see the design discussion continued in the issue; https://github.com/docker/docker/issues/24143#issuecomment-230650677, so it looks like this PR needs to be updated

/cc @aaronlehmann @aluzzardi ?

**Comment by johnharris85** [comment]
Ye @thaJeztah just waiting for the design stuff to be fleshed out over there before updating the PR.

File: api/client/swarm/opts.go [hunk 1]
```diff
@@ -112,6 +112,15 @@ func (o *AutoAcceptOption) Set(value string) error {
 			return fmt.Errorf("value NONE is incompatible with %s", value)
 		}
 		o.values[value] = true
+	case "ALL":
+		if accept, ok := o.values[WORKER]; ok && accept {
+			return fmt.Errorf("value NONE is incompatible with %s", WORKER)
```
Comment (aluzzardi) [inline api/client/swarm/opts.go]: This should just say `value NONE is compatible with ALL`

File: api/client/swarm/opts.go [hunk 2]
```diff
@@ -112,6 +112,15 @@ func (o *AutoAcceptOption) Set(value string) error {
...
+		}
+		if accept, ok := o.values[MANAGER]; ok && accept {
+			return fmt.Errorf("value NONE is incompatible with %s", MANAGER)
```
Comment (aluzzardi) [inline api/client/swarm/opts.go]: Same

Comment (johnharris85) [inline api/client/swarm/opts.go]: Thanks @aluzzardi, missed these but actually the logic needs to be tightened and can be made simpler now I think about it (after adding `ALL`). Just running tests on my changes and then will update the PR.

**Comment by johnharris85** [comment]
Added support for comma-separated values `--auto-accept worker,manager` as well as `all`. Also backwards compatible with the current way of setting these flags. Added tests to reflect the changes.

**Comment by cpuguy83** [comment]
I don't think `all` is a good idea.
Today we accept `manager`, `worker` and it's ok, but in the future we may have other types (`storage`?) where it makes less sense to support `all`

**Comment by cpuguy83** [comment]
meh, on second thought it's probably fine.

**Comment by johnharris85** [comment]
@cpuguy83 `all` is a UX feature for me. I think there is more refactoring to be done here (to make `all` future proof for example) but the current syntax `--auto-accept worker --auto-accept manager` is too long-hand IMO. The `--auto-accept manager,worker` syntax is not currently supported (this PR adds it). 

`all` makes even more sense when considering the potential for future options. If I'm doing some local experiments / development and I need to type `--auto-accept n` where _n_ is maybe 4 or 5 options. That's going to get old pretty quickly :p

**Comment by thaJeztah** [comment]
> `all` makes even more sense when considering the potential for future options.

I'm not sure we should develop for *possible* future options. What if the new future option is something that should be an "opt-in"? If someone set `all` (because: lazy), we could have a potential security issue. Better to keep it explicit, and have people update their config to allow the new type than to allow it without them knowing.

**Comment by johnharris85** [comment]
Maybe just allowing a comma-separated syntax is better then, and removing all? Would give a better UX shorthand instead of multiple `--auto-accept` flags but also preserve the security aspect?

**Comment by aaronlehmann** [comment]
@johnharris85: I think that's fine.

**Comment by thaJeztah** [comment]
Thanks! We can always add `all` back in future if we think it's ok; removing it is more troublesome once it's there

**Comment by johnharris85** [comment]
`all` removed and refactored some of the `none` checking logic to make it simpler. Tests amended.

File: api/client/swarm/opts.go [hunk 3]
```diff
@@ -18,6 +18,8 @@ const (
 	WORKER = "WORKER"
 	// MANAGER constant for manager name
 	MANAGER = "MANAGER"
+	// NONE constant
+	NONE = "NONE"
```
Comment (aaronlehmann) [inline api/client/swarm/opts.go]: Not strictly related to your change, but could you please change these constants to CamelCase? Identifiers should never be all-caps in Golang. I'm surprised golint doesn't complain about this.

https://golang.org/doc/effective_go.html#mixed-caps
https://github.com/golang/go/wiki/CodeReviewComments#mixed-caps

Comment (aaronlehmann) [inline api/client/swarm/opts.go]: Also, they should not be exported. And then we can drop the silly comments.

**Comment by aaronlehmann** [comment]
Can you please update the PR summary and description to reflect the changes from the review, and squash the commits?

File: api/client/swarm/opts.go [hunk 4]
```diff
@@ -99,27 +101,31 @@ func (o *AutoAcceptOption) String() string {
...
-	case WORKER, MANAGER:
-		if accept, ok := o.values[value]; ok && !accept {
-			return fmt.Errorf("value NONE is incompatible with %s", value)
+func (o *AutoAcceptOption) Set(acceptValues string) error {
+	for _, value := range strings.Split(acceptValues, ",") {
+		value = strings.ToUpper(value)
+		switch value {
+		case "", NONE:
+			// NONE must stand alone, so if any non-NONE setting exists, conflict
+			if !o.values[NONE] && len(o.values) > 0 {
```
Comment (aaronlehmann) [inline api/client/swarm/opts.go]: I'd suggest moving this check outside of the loop, with something like:

```go
if o.values[NONE] && len(o.values) > 1 {
        ...
}
```

Among other things, this will give a more consistent error message that won't depend on the order of the values.

File: api/client/swarm/opts.go [hunk 5]
```diff
@@ -99,27 +101,31 @@ func (o *AutoAcceptOption) String() string {
...
+				var s []string
+				for r := range o.values {
+					s = append(s, r)
+				}
+				return fmt.Errorf("value NONE is incompatible with %s", strings.Join(s, ", "))
+			}
+			o.values[WORKER] = false
+			o.values[MANAGER] = false
```
Comment (aaronlehmann) [inline api/client/swarm/opts.go]: This seems wrong because elsewhere we're looking at `len(o.values)`. Setting a value to false won't change the number of values in the map. We should be deleting the value instead.

The real problem is that the type of the map is `map[string]bool`. It should be `map[string]struct{}` instead, because this map is being used as a set. Making that change will prevent this kind of error.

Comment (aaronlehmann) [inline api/client/swarm/opts.go]: Actually there shouldn't be a need to remove these items anyway, because this is unreachable if either one was set to true. Also, removing them would interfere with moving the compatibility check out of the loop, as I suggested in https://github.com/docker/docker/pull/24146/files#r70158372.

I would still highly recommend making the type of `o.values` `map[string]struct{}`, though.

File: api/client/swarm/opts.go (hunk 3 above)
Comment (johnharris85) [inline api/client/swarm/opts.go]: They're being used in `opts_test.go`. So either they stay exported (but CamelCase) or we hard code the values into all the test cases?

Comment (aaronlehmann) [inline api/client/swarm/opts.go]: `opts_test.go` is in the same package, so they don't need to be exported.

File: api/client/swarm/opts.go (hunk 4 above)
Comment (johnharris85) [inline api/client/swarm/opts.go]: I actually don't see the value in specifying the explicit conflict, can't any error (where there is none + _n_) return with something like "value NONE cannot be specified alongside other node types" ?

File: api/client/swarm/opts.go (hunk 3 above)
Comment (johnharris85) [inline api/client/swarm/opts.go]: Right, thanks.

File: api/client/swarm/opts.go (hunk 4 above)
Comment (johnharris85) [inline api/client/swarm/opts.go]: Move to above line#105? `Set` is only called once though (and o.values is always empty until we set something in it) so that check will never eval to true right?

File: api/client/swarm/opts.go (hunk 5 above)
Comment (johnharris85) [inline api/client/swarm/opts.go]: Thanks for the feedback Aaron. Can you elaborate a little more on why this change would be more beneficial? (it was `map[string]bool` before I got to it, so just want to understand the benefits). Thanks!

File: api/client/swarm/opts.go (hunk 4 above)
Comment (aaronlehmann) [inline api/client/swarm/opts.go]: Agreed, I don't see the value in specifying the conflict.

I was thinking this could move below the loop, so it inspects `o.values` after the map is populated.

File: api/client/swarm/opts.go (hunk 5 above)
Comment (aaronlehmann) [inline api/client/swarm/opts.go]: `map[string]struct{}` is a map that has keys but no value (think "map of empty struct"). So it's a useful way to implement a set data structure. If you use `map[string]bool`, you have to think about two cases: either the key isn't present, or it is present with a value of `false`. If you use `map[string]struct{}`, there's no value, so the only consideration is which keys are present and which aren't. I find it a little less error-prone, and it also signals to experienced Go coders that the map is being used as a set.

There is also a slight efficiency improvement from not storing any values, but it wouldn't matter for something like this.

File: api/client/swarm/opts.go (hunk 4 above)
Comment (johnharris85) [inline api/client/swarm/opts.go]: Ah OK I get it. Ye can do, but with the standardizing of the error now does this have any other main benefits? I guess you could argue that error is checking the flag set as a whole, or that it's part of the `none` case. No strong feeling either way really.

Comment (aaronlehmann) [inline api/client/swarm/opts.go]: I think it makes most sense for it to be below the loop because otherwise you have to check for conflicts both in the `none` case (in case `none` comes after a different value) and in the other cases (in case they come after `none`). If the check is below the loop, only a single check is needed.

File: api/client/swarm/opts.go [hunk 6]
```diff
@@ -86,40 +88,35 @@ func NewListenAddrOption() NodeAddrOption {
...
+	values map[string]struct{}
 }
 
 // String prints a string representation of this option
 func (o *AutoAcceptOption) String() string {
 	keys := []string{}
-	for key, value := range o.values {
-		keys = append(keys, fmt.Sprintf("%s=%v", strings.ToLower(key), value))
+	for key := range o.values {
+		keys = append(keys, fmt.Sprintf("%s=%v", strings.ToLower(key), true))
```
Comment (aaronlehmann) [inline api/client/swarm/opts.go]: Just noticed this: should this be a list of values, instead of `x=true, y=true`? There is no `=false`, so the `=true` seems redundant.

Comment (johnharris85) [inline api/client/swarm/opts.go]: Ye I thought that too, maybe just for readability / clarity (for repr), although ye, not strictly required as it's true if it exists by default. Happy to change but think it's clearer to stay. No strong preference either way though.

File: api/client/swarm/opts.go [hunk 7]
```diff
@@ -133,17 +130,32 @@ func (o *AutoAcceptOption) Policies(secret *string) []swarm.Policy {
...
 }
 
+// IsPresent returns whether the key exists in the set or not
+func (o *AutoAcceptOption) IsPresent(key string) bool {
+	_, c := o.values[key]
+	return c
+}
+
+// AddValue adds a value to the set of auto-accepted nodes
+func (o *AutoAcceptOption) AddValue(value string) {
```
Comment (aaronlehmann) [inline api/client/swarm/opts.go]: Please unexport `IsPresent` and `AddValue` since they aren't meant to be used outside the package.

You might also consider removing `AddValue` and doing the map key assignment directly.

File: api/client/swarm/opts.go [hunk 8]
```diff
@@ -30,9 +30,11 @@ const (
 
 var (
 	defaultPolicies = []swarm.Policy{
-		{Role: WORKER, Autoaccept: true},
-		{Role: MANAGER, Autoaccept: false},
+		{Role: worker, Autoaccept: true},
+		{Role: manager, Autoaccept: false},
 	}
+
+	exists = struct{}{}
```
Comment (aaronlehmann) [inline api/client/swarm/opts.go]: You can just use `struct{}{}` directly instead of defining a variable for it. It doesn't matter either way, but that's more typical usage.

File: api/client/swarm/opts.go (hunk 6 above)
Comment (aaronlehmann) [inline api/client/swarm/opts.go]: Okay, if you decide to keep the current format, can you change the Sprintf call to `fmt.Sprintf("%s=true", strings.ToLower(key))`?

File: api/client/swarm/opts.go (hunk 8 above)
Comment (LK4D4) [inline api/client/swarm/opts.go]: I wonder what the point of using `struct{}` instead of bool. `bool` syntax is much simpler.

Comment (aaronlehmann) [inline api/client/swarm/opts.go]: @lk4d4 Check out the comments on outdated parts of the diff in this PR.

Comment (LK4D4) [inline api/client/swarm/opts.go]: @aaronlehmann heh, okay. I still think that readability of bool version is much better and I wouldn't prefer `struct{}` version for one-time set usage.

File: api/client/swarm/opts.go [hunk 9]
```diff
@@ -86,40 +86,35 @@ func NewListenAddrOption() NodeAddrOption {
...
-		o.values[WORKER] = false
-		o.values[MANAGER] = false
-	case WORKER, MANAGER:
-		if accept, ok := o.values[value]; ok && !accept {
-			return fmt.Errorf("value NONE is incompatible with %s", value)
+func (o *AutoAcceptOption) Set(acceptValues string) error {
+	for _, value := range strings.Split(acceptValues, ",") {
+		value = strings.ToUpper(value)
+		switch value {
+		case "", none:
```
Comment (aaronlehmann) [inline api/client/swarm/opts.go]: Sorry, one final comment. I'm wondering it makes sense to treat `""` the same as `none`. It would mean that `worker,,manager` is parsed as `worker,none,manager`. I guess the intent here is to treat `--auto-accept ""` as `--auto-accept none`. If that's the case, maybe it would be better to do it at the very beginning of this function:

```go
if acceptValues == "" {
        o.values[none] = struct{}{}
} else {
        for _, value := range strings.Split(acceptValues, ",") {
                ...
        }
}
```

Comment (johnharris85) [inline api/client/swarm/opts.go]: Good point, but then `worker,,manager` would trigger the `default` case? In addition to the above change, should a case be added for an empty string inside the loop to just ignore it?

Comment (aaronlehmann) [inline api/client/swarm/opts.go]: I'm not sure, but either sounds reasonable to me. Triggering an error in that case (by hitting the default case) is probably the safest approach.

Comment (johnharris85) [inline api/client/swarm/opts.go]: And ye, I guess that was the original intent (that code existed before my PR) but I'm not sure whether I agree with it. I think a blank `--auto-accept` should error or pass silently, not set to `none`. That doesn't seem intuitive behaviour. Perhaps it was done for security reasons? In case someone set it accidentally but didn't pass anything, default to `none`. That may have made sense before, but aren't the defaults changing to `false` for both manager and worker anyway?

Comment (aaronlehmann) [inline api/client/swarm/opts.go]: > And ye, I guess that was the original intent (that code existed before my PR) but I'm not sure whether I agree with it. I think a blank --auto-accept should error or pass silently, not set to none.

Makes sense.

> That doesn't seem intuitive behaviour. Perhaps it was done for security reasons? In case someone set it accidentally but didn't pass anything, default to none.

Not really sure that makes sense. It's hard to accidentally pass an empty string as an argument.

> That may have made sense before, but aren't the defaults changing to false for both manager and worker anyway?

The autoaccept defaults aren't changing. What changed recently on master is that by default, a secret token is needed to join a swarm.

**Comment by aaronlehmann** [comment]
Code LGTM

Can you please update the commit message and pull request description to summarize the final state of the changes?

This will also need updates to the docs before merging.

Thanks!

**Comment by cpuguy83** [comment]
LGTM

**Comment by cpuguy83** [comment]
ping @thaJeztah

**Comment by thaJeztah** [comment]
Thanks! docs LGTM

File: docs/reference/commandline/swarm_init.md [hunk 10]
```diff
@@ -66,6 +66,13 @@ For example, the following initializes a cluster with auto-acceptance of workers
 $ docker swarm init --listen-addr 192.168.99.121:2377 --auto-accept worker
 ```
 
+It is possible to pass a comma-separated list of node types. The following initializes a cluster
+with auto-acceptance of both `worker` and `manager` nodes
```
Comment (aaronlehmann) [inline docs/reference/commandline/swarm_init.md]: I think `none` needs to be mentioned somewhere in the docs.

Comment (thaJeztah) [inline docs/reference/commandline/swarm_init.md]: @aaronlehmann good point; just created https://github.com/docker/docker/pull/24552

**Comment by aluzzardi** [comment]
Ping @icecrime
